File: structy/structy-js/Array-and-String/mostFrequentChar.js
// most frequent char
// Write a function, mostFrequentChar, that takes in a string as an argument. The function should return the most frequent character of the string. If there are ties, return the character that appears earlier in the string.

// You can assume that the input string is non-empty.


const mostFrequentChar = (s) => {
	const count = {};

	for (let char of s) {
		// console.log(char);
		if (char in count) {
			count[char]++;
		} else {
			count[char] = 1;
		}
		// console.log(count);
	}

	let best;
	for (let char of s) {
		// console.log('value: ', count[char]);
		// console.log('best: ', best);
		if (!best || count[best] < count[char]) {
			best = char;
		}
	}

	return best;
};

console.log(mostFrequentChar('bookeeper')); // -> 'e'
console.log(mostFrequentChar('david')); // -> 'd'
console.log(mostFrequentChar('abby')); // -> 'b'
console.log(mostFrequentChar('mississippi')); // -> 'i'
console.log(mostFrequentChar('potato')); // -> 'o'
console.log(mostFrequentChar('eleventennine')); // -> 'e'
console.log(mostFrequentChar('riverbed')); // -> 'r'

File: structy/structy-py/1-array_and_string/most_frequent_char.py
# Write a function, most_frequent_char, that takes in a string as an argument. The function should return the most frequent character of the string. If there are ties, return the character that appears earlier in the string.

# You can assume that the input string is non-empty.

def most_frequent_char(s):
    count = {}
    for char in s:
        if char not in count:
            count[char] = 0
        count[char] += 1

    most_freq = None
    for char in s:
        if most_freq is None or count[char] > count[most_freq]:
            most_freq = char
    return most_freq


##!TEST_00
print(most_frequent_char('bookeeper'))  # -> 'e'

##!TEST_00
print(most_frequent_char('david'))  # -> 'd'

##!TEST_00
print(most_frequent_char('abby'))  # -> 'b'

##!TEST_00
print(most_frequent_char('mississippi'))  # -> 'i'

##!TEST_00
print(most_frequent_char('potato'))  # -> 'o'

##!TEST_00
print(most_frequent_char('eleventennine'))  # -> 'e'

##!TEST_00
print(most_frequent_char('riverbed'))  # -> 'r'
